pytools/graph.py

```diff
@@ -240,6 +241,42 @@ def __init__(self, node: NodeT) -> None:
         self.node = node
 
 
+def find_cycles(graph: GraphT) -> List[List[NodeT]]:
+    """
+    Find all cycles in *graph* using DFS.
+
+    :returns: A :class:`list` in which each element represents another :class:`list`
+        of nodes that form a cycle.
+    """
+    def dfs(node: NodeT, path: List[NodeT]) -> List[NodeT]:
```

Comment:
Constructing `path` just in case is wasteful IMO: The path could be collected as you return, if a cycle is found.

Reply:
What do you think of [96bd3a6](https://github.com/inducer/pytools/pull/167/commits/96bd3a66e13aca8adcdb8cdee63965dd2bba15f9)?